src/Interpreters/Cache/FileCache.cpp

```diff
@@ -817,6 +823,69 @@ bool FileCache::tryReserve(FileSegment & file_segment, const size_t size, FileCa
     return true;
 }
 
+bool FileCache::isFreeSpaceRatioSatisfied(const CacheGuard::Lock & lock) const
+{
+    size_t size_limit = main_priority->getSizeLimit();
+    size_t elements_limit = main_priority->getElementsLimit();
+
+    return (size_limit == 0
+            || main_priority->getSize(lock) < keep_current_size_to_max_ratio * size_limit)
+        && (elements_limit == 0
+            || main_priority->getElementsCount(lock) < keep_current_elements_to_max_ratio * elements_limit);
+}
+
+void FileCache::freeSpaceRatioKeepingThreadFunc()
+{
+    while (true)
+    {
+        auto lock = lockCache();
+        if (shutdown_called)
+            return;
+
+        try
+        {
+            if (isFreeSpaceRatioSatisfied(lock))
+            {
+                cache_evicting_cv.wait(lock, [&](){ return shutdown_called || !isFreeSpaceRatioSatisfied(lock); });
+                if (shutdown_called)
+                    return;
+            }
+
+            keepUpFreeSpaceRatio(lock);
```

Comment:
The background thread shouldn't try to achieve the desired ratio at once, it should do it in some chunks (and do some sleeps between them; otherwise, theoretically, it can take the global lock over and over w/o giving not much chance to other threads. Probably, @serxa can enlighten us regarding such possibilities). Otherwise, `tryReserve()` and queries, which read from cache, can wait much longer than necessary. So, logically, the background thread should have "less priority" than the threads which do the work on demand.

Reply:
yes, also thought about this